compiler/rustc_data_structures/src/obligation_forest/mod.rs

```diff
@@ -424,7 +431,10 @@ impl<O: ForestObligation> ObligationForest<O> {
             // `for index in 0..self.nodes.len() { ... }` because the range would
             // be computed with the initial length, and we would miss the appended
             // nodes. Therefore we use a `while` loop.
-            let mut index = 0;
+
+            let mut index =
+                processor.skippable_obligations(self.nodes.iter().map(|n| &n.obligation));
+
```

Comment:
You said on Zulip you are seeing big speedups for `keccak`, which is great! But I don't understand how. This preprocessing step doesn't seem any cheaper than the old code. Determining how many items to skip seems to take as much work as just processing them in the loop?

Reply:
* it only does the RefCell borrow once, that way there are no memory-writes inside the loop, only reads
* in `ena` there's no cold path to update the value, instead it just bails
* no undo log in the infer system
* all the matches only have 2 cases (success/failure) now.

Overall that makes the loop much simpler and I assume a lot easier for the CPU to speculate